### PR TITLE
issue #242 implementation of repNM for 1.2.x

### PR DESCRIPTION
**Pull request by pgrandjean**

### CODE REVIEW

**Comment by pgrandjean** [comment]
I have signed the Scala CLA.

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala [hunk 1]
```diff
@@ -792,6 +792,57 @@ trait Parsers {
       applyp(in)
     }
 
+  /** A parser generator for a specified range of repetitions interleaved by a
+   * separator.
+   *
+   *  `repN(n, m, p, s)` uses `p` at least `m` times and up to `n` times, interleaved
```
Comment (Philippus) [inline shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]:
```suggestion
   *  `repN(m, n, p, s)` uses `p` at least `m` times and up to `n` times, interleaved
```

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala [hunk 2]
```diff
@@ -792,6 +792,57 @@ trait Parsers {
...
+
+    mandatory(in) match {
+      case Success(x, rest) => elems ++= x; continue(rest)
+      case ns: NoSuccess => ns
+    }
+  }
+
+  /** A parser generator for a specified range of repetitions.
+   *
+   *  `repN(n, m, p)` uses `p` at least `m` times and up to `n` times to parse the input
```
Comment (Philippus) [inline shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]:
```suggestion
   *  `repN(m, n, p)` uses `p` at least `m` times and up to `n` times to parse the input
```

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala [hunk 3]
```diff
@@ -792,6 +792,57 @@ trait Parsers {
...
+   * @param m   minimum number of repetitions
+   * @param n   maximum number of repetitions
+   * @param p   a `Parser` that is to be applied successively to the input
+   * @param sep a `Parser` that interleaves with p
+   * @return    A parser that returns a list of results produced by repeatedly applying `p` interleaved
+   *            with `sep` to the input. The list has a size between `m` and up to `n`
+   *            (and that only succeeds if `p` matches at least `n` times).
+   */
+  def repMN[T](m: Int, n: Int, p: Parser[T], sep: Parser[Any]): Parser[List[T]] = Parser { in =>
+    require(0 <= m && m <= n)
```
Comment (Philippus) [inline shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: Can we do without the `require`?

**Comment by SethTisue** [comment]
fwiw, https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html uses the "n, m" ordering (not "m, n")

**Comment by SethTisue** [comment]
I guess the parameters need names, but baking the NM/MN thing into the name of the trait (`repMN`) isn't necessary, I think? if we have `repN`, then the other one ought to be `repNM`, I think, not `repMN`. but perhaps we could dodge the issue (except in the parameter names) with a name like `repRange`... or something?

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala (hunk 3 above)
Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: removed

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala (hunk 2 above)
Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: as suggested by @SethTisue, I renamed repMN into repNM to stay in line with java's regex Pattern.

**Comment by pgrandjean** [comment]
@SethTisue I like repNM, it's pretty compact and intuitive if you know regex.

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala (hunk 1 above)
Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: fixed.

Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: following @SethTisue suggestion, I renamed the function to `repNM`.

File: shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala [hunk 4]
```diff
@@ -792,6 +792,56 @@ trait Parsers {
...
+   *  (the result is a `List` of at least `n` consecutive results of `p` and up to `m` results).
+   *
+   * @param n   minimum number of repetitions
+   * @param m   maximum number of repetitions
+   * @param p   a `Parser` that is to be applied successively to the input
+   * @return    A parser that returns a list of results produced by repeatedly applying `p` to the input.
+   *            The list has a size between `n` and up to `m`
+   *            (and that only succeeds if `p` matches at least `n` times).
+   */
+  def repNM[T](n: Int, m: Int, p: Parser[T]): Parser[List[T]] = repNM[T](n, m, p, success(()))
```
Comment (SethTisue) [inline shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: could/should this be a default argument, instead of using overloading?

Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: yes, you're right.

Comment (pgrandjean) [reply shared/src/main/scala/scala/util/parsing/combinator/Parsers.scala]: fixed

**Comment by SethTisue** [comment]
@Philippus this seems ready for merge once Travis-CI likes it; I think it's failing currently because this fix is needed: https://github.com/scala/scala/pull/8536

**Comment by lrytz** [comment]
Cleared the travis caches and re-started the build, it's green now.

**Comment by pgrandjean** [comment]
@SethTisue @Philippus @lrytz thank you!

**Pull request merged**